### PR TITLE
feat: add support for cloud files and non-local checkpoints

### LINKED ISSUE #19412: S3 support for WandBLogger
### Description & Motivation

It would be great to have direct s3 support with the `WandBLogger`. 

e.g. specifying
```python
logger = WandBLogger(save_dir="s3://{bucket}/{path}")
cli = Trainer(logger=logger)
...
``` 
Will automatically log WandB artifacts to S3 bucket!

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by parambharat**
## What does this PR do?
This PR adds support to log model checkpoints on S3 and other cloud file systems as artifact references to Weights & Biases.

Fixes #19412 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20701.org.readthedocs.build/en/20701/

### CODE REVIEW

**Comment by parambharat** [comment]
Hi @awaelchli . I have created this PR to address the issue #19412 . However, I'm having difficulty with writing the unit tests. Specifically, with mocking the remote file systems. Could you point me to where the tests for the [model checkpoint callback](https://github.com/Lightning-AI/pytorch-lightning/blob/831870a15a17cca4152ffde5c8a3ebc535c68ab0/src/lightning/pytorch/callbacks/model_checkpoint.py#L50) checks remote file systems ? I'll try to emulate those tests in mocking remote filesystems.
